ansible/roles/infra-osp-resources-destroy/tasks/delete_osp_resources.yml

```diff
@@ -30,19 +30,36 @@
             openstack project purge --keep-project --project {{ osp_project_name }}
           when: project_purge_out is succeeded
 
-        - name: Delete any remaining volumes in project
-          shell: >-
-            set -o pipefail;
-            openstack volume list --project {{ osp_project_name }} -f value -c ID
-            | xargs openstack volume delete
-        
-        - name: Delete any trunk ports in project
-          shell: >-
-            set -o pipefail;
-            openstack port list --project {{ osp_project_name }} -f json -c trunk_details
-            | jq '.[].trunk_details.trunk_id | select (.!=null)'
-            | xargs openstack network trunk delete
-        
+        - name: Get all remaining volumes in project
```

Comment:
I'm really confused why we turned 12 lines of code with 2 commands into over double the amount for this? This seems like overcomplicating what can be done much simpler. If there are volumes or trunks to delete, it deletes them. If not, it moves on to the remainder.

is there a reason we need the information collected, added as facts, and then essentially running a very similar delete command?

Reply:
I got errors when trying ocp-clientvm with that "shell" tasks.

The first problem was that when no volume was found, the task was failing

```
TASK [infra-osp-resources-destroy : Delete any remaining volumes in project] ******************************************************************************************************************
Friday 22 November 2019  07:53:42 +0000 (0:00:12.699)       0:00:24.214 ******* 
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "set -o pipefail; openstack volume list --project guid01-project -f value -c ID | xargs openstack volume delete", "delta": "0:00:05.741398", "end": "2019-11-22 07:53:48.226543", "msg": "non-zero return code", "rc": 123, "start": "2019-11-22 07:53:42.485145", "stderr": "usage: openstack volume delete [-h] [--force | --purge]\n                               <volume> [<volume> ...]\nopenstack volume delete: error: the following arguments are required: <volume>", "stderr_lines": ["usage: openstack volume delete [-h] [--force | --purge]", "                               <volume> [<volume> ...]", "openstack volume delete: error: the following arguments are required: <volume>"], "stdout": "", "stdout_lines": []}
```


The second error was:

```

TASK [infra-osp-resources-destroy : Delete any trunk ports in project] ************************************************************************************************************************
Friday 22 November 2019  12:30:46 +0000 (0:00:00.021)       0:00:19.720 ******* 
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "set -o pipefail; openstack port list --project guid01-project -f json -c trunk_details | jq '.[].trunk_details.trunk_id | select (.!=null)' | xargs openstack network trunk delete", "delta": "0:00:03.332714", "end": "2019-11-22 12:30:50.296071", "msg": "non-zero return code", "rc": 123, "start": "2019-11-22 12:30:46.963357", "stderr": "/bin/sh: jq: not found\nusage: openstack network trunk delete [-h] <trunk> [<trunk> ...]\nopenstack network trunk delete: error: the following arguments are required: <trunk>\nException ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>\nBrokenPipeError: [Errno 32] Broken pipe", "stderr_lines": ["/bin/sh: jq: not found", "usage: openstack network trunk delete [-h] <trunk> [<trunk> ...]", "openstack network trunk delete: error: the following arguments are required: <trunk>", "Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>", "BrokenPipeError: [Errno 32] Broken pipe"], "stdout": "", "stdout_lines": []}
```

`jq` binary was not available in my docker images.  I can easily add it, but it's also a new dependency for anybody deploying from their laptop since this is running from localhost.

It's difficult to think of everything when using a `shell` task, spliting into separate tasks is good practice.

You noticed i added:

```bash
set -o pipefail
```

Tells the shell to fail if any command from the pipe fails.
we want to know if `openstack volume list` failed. Because if it fails silently, then we can't be sure we deleted everything.

But you have to be sure of every return code of every commands in the pipe.
And they are not always consistent.

On the other hand, you don't necessary want to remove `set -o pipefail`, you want to know if you "read" command failed. So for something apparently very easy, it quickly become a puzzle.

When you split the commands into separate tasks, it gives you flexibility.

You can even then start defining:

- `failed_when`  (using rc and what not)
- `changed_when` (read-only tasks should have `changed_when: false`
- rescue block... 
- etc

It's more "lines of code" but it's more robust and flexible.

Reply:
We'll have to agree to disagree on some of this since I think there is a happy medium between simplicity and robustness, but I accept your changes and arguments to make this work as needed (even if adding a little complexity!).